In [ ]:
pip install spacy -q   # -q hides output

In [ ]:
!python -m spacy download nl_core_news_sm -q

/usr/local/lib/python3.8/dist-packages/torch/cuda/__init__.py:497: UserWarning: Can't initialize NVML
  warnings.warn("Can't initialize NVML")
2023-01-21 21:05:32.254772: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 92.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('nl_core_news_sm')


In [ ]:
pip install spacy-transformers -q

In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
import sys
sys.path.append('/content/drive/My Drive/DTA/Internship/')

# Process Data

- Convert .csv format data into .spacy data
- Train/test/dev split
- Save files

In [ ]:
import spacy
from transform_split_save_spacy import make_train_data, training_to_docbin

In [ ]:
train_data, test_data, dev_data = make_train_data("drive/MyDrive/DTA/Internship/C14NL.csv")

train_docbin = training_to_docbin(train_data)
test_docbin = training_to_docbin(test_data)
dev_docbin = training_to_docbin(dev_data)

data_filename = "drive/MyDrive/DTA/Internship/spacy_data"

train_docbin.to_disk(data_filename + "_TRAIN.spacy")
test_docbin.to_disk(data_filename + "_TEST.spacy")
dev_docbin.to_disk(data_filename + "_DEV.spacy")

100%|██████████| 124/124 [00:00<00:00, 416.25it/s]

[sotteghem, laureins, van, lancackere, katelinen, sputters, twintich, scellinghe, par, ghent, kersauonde, sente, marien, licht, messe, sotteghem, katteline, sputters, laurense, meyers, pieters, ver, minnen, alsbrouc, martine, ghisels, kateline, twintich, scellinghe, par, laureinse, laureins, kateline, tsente, marien, licht, messe, laureins, twalef, penninghe, parisise, dertienhondert, ende, achte, ende, veertich, vp, sente, marcs, dach, jn, aprille]
[zottenghem, jan, van, der, boenbeken, pouwelse, van, den, dale, tiene, scellinghe, parisise, alre, heyleghen, messe, sente, martijns, messe, tien, scellinghe, parisise, pouwels, janne, claus, baertmakers, claus, van, nederhem, pouwels, tien, scellinghe, parisise, janne, jan, jan, pouwels, tien, scellinghe, parisise, drie, scellinghe, parisise, pouwelse, dertienhondert, zesse, ende, viftich, den, zeuentienden, dach, van, laumaende]
[gillis, de, meyere, mergrieten, gillis, gadmins, pamele, schelde, hamme, de, tichelrie, mgrieten, pamele, jac


100%|██████████| 15/15 [00:00<00:00, 254.63it/s]


[jan, van, den, keye, vinhouters, zeghelsem, onser, vrouwen, audenaerde, iij, lb, parisise, xxx, sc, par, kersauonde, sente, marien, lichtmesse, kersauonde, xxx, sc, par, sente, jans, messe, midden, somere, sente, pieters, daghe, oust, iij, lb, par, xxx, sc, par, kersauonde, daniels, van, den, eechoute, jans, van, den, eechoute, zeghelsem, eechoute, lobis, jan, van, den, eechoute, xxx, sc, par, kersauonde, xxx, sc, par, sente, jans, messe, midden, somere, xxx, sc, par, sente, jans, messe, midden, somere, lobis, jan, van, den, keye, vinhouters, marie, cabboets, onser, vrouwen, auden, marie, cabboets, marien, cabboets, scoeresse, ghiselbrechte, den, meyere, willeme, heeren, daniel, de, neve, gillis, van, den, keye, gillis, screuel, willem, van, torout, jan, de, weer, zeghelsem, mccc, en, xlix, sfriendachs, nar, sente, niclaus, daghe]
[kateline, jane, van, zeghelsem, zeghelsem, onser, vrauwen, audenaerde, xv, sc, par, kersauonde, sente, marien, lichtmesse, her, mijns, zeghelsem, lobis, ka

100%|██████████| 16/16 [00:00<00:00, 345.06it/s]

[sottenghem, jan, de, brakelare, sente, gorics, oudenhoue, gosine, van, d, swalmen, sente, martijns, messe, kersthauonde, sottenghe, gosijn, jan, zesse, en, twintich, schellinghe, par, jan, de, brakelare, gosine, willem, tants, jan, zesse, en, twintich, schellinghe, par, gosine, gosijn, dertienhondert, neghen, en, viertich, tsondaechs, na, sentecatelinen, dach]
[zotteghem, lambrecht, brandijn, ysendike, janne, den, smet, van, den, bossche, van, den, eeckoute, viere, ende, twintich, scellinghe, par, kerssauonde, onser, vrouwen, daghe, ter, lichtmessen, viere, ende, twintich, scellinghe, par, jan, lambrechte, ghiiskens, jan, braemboschs, zigghers, metten, jan, lambrechte, lambrecht, viere, scellinghe, par, dertienhondert, ende, drie, ende, zestich, den, tienden, dach, van, meye]
[pauwels, van, detome, clays, stalin, willem, alnaect, michiel, van, de, berghe, danin, de, vos, eluerdinghe, vlaemertinghe, nicasis, de, vroede, jehan, van, de, clite, lambrecht, voets, lotin, balghe, vlamtinghe

# Make config file

### Option 1:

Make and download base_config.cfg from the spaCy quickstart widget: https://spacy.io/usage/training#quickstart

Fill config file with command:

> python -m spacy init fill-config base_path output_file --diff

Example: 
> python -m spacy init fill-config base_config.cfg config.cfg --diff 

base_config.cfg should be the file you just made and downloaded from the quickstart widget

--diff = prints the differences between base_config and config

### Option 2:

Make and fill config file with one command:
> python -m spacy init config output_file --lang --pipeline --optimize --gpu

Example:
> python -m spacy init config config.cfg -l nl -p ner -o accuracy --gpu

- **-l** = "--lang", we set our language to Dutch (nl)
- **-p** = "--pipeline", we set our pipeline components to only NER
- **-o** = "--optimize", we optimize for accuracy instead of efficiency
- **--gpu** = we enable GPU



More info: https://spacy.io/api/cli#init

# Debug data

In order to debug the data, or check for problems, we first have to manually add our train and dev data paths to the config.cfg file.

This can be done in the very first lines of the file:
> [paths]

> train = path_to_your_data_file/spacy_data_TRAIN.spacy

> dev = path_to_your_data_file/spacy_data_DEV.spacy

Now we can use the command:

> python -m spacy debug data config.cfg

In our case, we have one warning: "Low number of examples to train a new pipeline (124)"

Our data has only 155 charters in total, the training data is only 124.


# Training

Assuming the model will be trained on google colab, you can either replace the paths in the previous step with the corresponding paths in your google drive and use the first command, or leave them blank ("null") and use the second command:

If you filled in the train and dev paths:
> python -m spacy train config_path --output --gpu-id 


- **--output** = directory to output trained models in
- **gpu-id** set to 0 to enable

If you left the paths blank:
> python -m spacy train config.cfg --output your_output_path --paths.train your_train_path --paths.dev your_dev_path --gpu-id 0

Example in Google Colab:
> pip install spacy

> pip install spacy-transformers

> python -m spacy train config.cfg --output drive/MyDrive/spaCy/output --paths.train drive/MyDrive/spaCy/training_data_TRAIN.spacy --paths.dev drive/MyDrive/spaCy/training_data_DEV.spacy --gpu-id 0

### Output interpretation

During training he following information will be displayed:

E = epoch

\# = steps/batches

LOSS TRANSFORMER = loss for transformer pipeline component

LOSS NER = loss for NER pipeline component

ENTS_F = f-score for entities

ENTS_P = precision for entities

ENTS_R = recall for entities

SCORE = overall accuracy

# Testing model

After the training is done, we can load the best model and use it.

In [ ]:
nlp_ner = spacy.load("drive/MyDrive/spaCy/output/model-best")

In [ ]:
colors = {"B-PERS": "#F67DE3", "I-PERS": "#F9C0F0", "B-LOC":"#26DDB1", "I-LOC":"#85F9DD", "B-DATE":"#427AF9", "I-DATE":"#ABC5FF", "B-MONEY":"#FFF11E", "I-MONEY":"#FCF7A8"}
options = {"colors": colors} 

In [ ]:
text = "wje lambrecht de bere johan van breinen stacin de vos clais cruke en pieter stacin scepenen mire vrouwen der abbedessen van messine jn zuutscoten en jn noorscoten jn dien tiden doen te weitene allen den ghonen die dese lettre zullen zien of horen lesen dat fransois croeselin portre jn ypre heift ghecocht en ghecreghe te sinen bouf en te sijns hoirs bouf jegh masine van den walle portre jn ypre viere ymeite twe linen en ene halue en twintich roeden lants litel min of litel meer ligghende binnen der proghie van noorscoten namenachtelike binnen der vierscare voors dats te weitene jn al zulken paertselen als hier naer volghet al voren jn poesele jn den oost brouc viere linen en veertich roeden litel min of meer tussche fransois croeselins lande an dene zide en der wedewe en der aeldinghe johans waghenaers lande an dander zide jtem een half ymet litel min of meer jn den calewaert tussche fransois croeselins lande an dene zide en sdisch van noorscoten lande an dander zide jte een ymet en viue roeden litel min of meer jn de noort deghede tussche fransois croselins lande an den zide en stacin hein sons lande an dander zide jt een half ymet litel min of meer tussche fansois croeselins lande an dene zide en johans baens lande an dander zide en ute al desen voors lande gaet tiene scellinghe en zeuen pene par sjaers erueliker rente den couente van messine jte een half ymet en ertich roeden west waert ouer de ypre jn der aeldinghe sense braems lande jte twe linen en ene halue benoorden jan lauwers tussche franse buulscamps lande an dene zide en johan lauwers lande an dander zide ende ute desen tween linen en ene halue gaet zeuentiene scellinghe en zesse pene par sjaers erueliker rente den canoneken van messine jte so heift die voors fransois croeselin ghecocht jegh den voors masine van den walle viere scelinghe en tiene pene pars en een hallinc sjaers erueliker renten de welke sijn beset up de steide die kerstiaen vels was van al den welken voors lande en rente die voors masin van denwalle liet uulgouden en wel betaelt en heuets ghegheuen halm en ghifte bede van den lande en van der rente den voors frasois croeseline wel en wettelike naer costumen en vsagen van der vierscare voors en heueets hem gheloft ghewaersceipe te warendeirne van allen calaengen jeghe allen lieden behouden der rente die hier voors es en voort quite lant omme quite ghelt behouden mire vrouwen lantscout diereute gaet jn orconsceipen van al den voors dinghe en bi versoukene van beden partien so hebben wie scepene voors dese lettre gheseghelt met onsen zeghelen uuthanghende dit was ghedaen jnt jaer ons heren als men screef mccc achte en vichtich up den viere en twintichsten dach van maerte'"

In [ ]:
doc = nlp_ner(text)

spacy.displacy.render(doc, style="ent", options= options, jupyter=True)

# Package Pipeline



Next, we want to package our pipeline so we can upload it to Huggingface. For this, it was necessary (in my case) to make an Anaconda environment with Python 3.10 and Spacy version 3.5.

Install spacy-huggingface-hub:
> pip install spacy-huggingface-hub

These installations will possibly be necessary too:
> conda install huggingface_hub

> pip install chardet

### IMPORTANT: 
- The model name for Huggingface must be assigned NOW (in the following command)!! If it is not assigned now, changing the name at any point later than this will cause problems.
- The model name must have NO dashes (-), only underscores (_)

> python -m spacy package model_best_path output_dir --build wheel –name pipeline_name 

Example:
> python -m spacy package model-best output --build wheel –name pipeline_name

# Log in to Hugginface

In order to upload our model to Huggingface, we must first log in to Huggingface in the command line:

Enter the following command:
> huggingface-cli login

You will be prompted to input your user authentication token. This can be found in your profile on Huggingface.

You will also be advised to run the following command in case you want to set this credential helper as the default:

> git config --global credential.helper store

# Push to Huggingface

To push your model to Huggingface, use the following command: 

> python -m spacy huggingface-hub push pipeline_name-0.0.0-py3-none-any.whl

pipeline_name will be the name you entered as pipeline name when packaging the pipeline. It seems the language of your model will automatically be added before the name of your pipeline. In our case, the name we used when packaging was "middle_dutch_spacy", it was changed to "nl_middle_dutch_spacy".

# Use Huggingface model

In [ ]:
!pip install https://huggingface.co/Alassea/nl_middle_dutch_spacy/resolve/main/nl_middle_dutch_spacy-any-py3-none-any.whl -q

# Using spacy.load().
import spacy
nlp = spacy.load("nl_middle_dutch_spacy")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 406.1/406.1 MB 3.5 MB/s eta 0:00:00


In [ ]:
colors = {"B-PERS": "#F67DE3", "I-PERS": "#F9C0F0", "B-LOC":"#26DDB1", "I-LOC":"#85F9DD", "B-DATE":"#427AF9", "I-DATE":"#ABC5FF", "B-MONEY":"#FFF11E", "I-MONEY":"#FCF7A8"}
options = {"colors": colors} 

In [ ]:
text = "wje lambrecht de bere johan van breinen stacin de vos clais cruke en pieter stacin scepenen mire vrouwen der abbedessen van messine jn zuutscoten en jn noorscoten jn dien tiden doen te weitene allen den ghonen die dese lettre zullen zien of horen lesen dat fransois croeselin portre jn ypre heift ghecocht en ghecreghe te sinen bouf en te sijns hoirs bouf jegh masine van den walle portre jn ypre viere ymeite twe linen en ene halue en twintich roeden lants litel min of litel meer ligghende binnen der proghie van noorscoten namenachtelike binnen der vierscare voors dats te weitene jn al zulken paertselen als hier naer volghet al voren jn poesele jn den oost brouc viere linen en veertich roeden litel min of meer tussche fransois croeselins lande an dene zide en der wedewe en der aeldinghe johans waghenaers lande an dander zide jtem een half ymet litel min of meer jn den calewaert tussche fransois croeselins lande an dene zide en sdisch van noorscoten lande an dander zide jte een ymet en viue roeden litel min of meer jn de noort deghede tussche fransois croselins lande an den zide en stacin hein sons lande an dander zide jt een half ymet litel min of meer tussche fansois croeselins lande an dene zide en johans baens lande an dander zide en ute al desen voors lande gaet tiene scellinghe en zeuen pene par sjaers erueliker rente den couente van messine jte een half ymet en ertich roeden west waert ouer de ypre jn der aeldinghe sense braems lande jte twe linen en ene halue benoorden jan lauwers tussche franse buulscamps lande an dene zide en johan lauwers lande an dander zide ende ute desen tween linen en ene halue gaet zeuentiene scellinghe en zesse pene par sjaers erueliker rente den canoneken van messine jte so heift die voors fransois croeselin ghecocht jegh den voors masine van den walle viere scelinghe en tiene pene pars en een hallinc sjaers erueliker renten de welke sijn beset up de steide die kerstiaen vels was van al den welken voors lande en rente die voors masin van denwalle liet uulgouden en wel betaelt en heuets ghegheuen halm en ghifte bede van den lande en van der rente den voors frasois croeseline wel en wettelike naer costumen en vsagen van der vierscare voors en heueets hem gheloft ghewaersceipe te warendeirne van allen calaengen jeghe allen lieden behouden der rente die hier voors es en voort quite lant omme quite ghelt behouden mire vrouwen lantscout diereute gaet jn orconsceipen van al den voors dinghe en bi versoukene van beden partien so hebben wie scepene voors dese lettre gheseghelt met onsen zeghelen uuthanghende dit was ghedaen jnt jaer ons heren als men screef mccc achte en vichtich up den viere en twintichsten dach van maerte'"

In [ ]:
nlp(text)

spacy.displacy.render(doc, style="ent", options= options, jupyter=True)